diff --git a/src/JITModule.cpp b/src/JITModule.cpp
index 0a50d08e3..274d9b223 100644
--- a/src/JITModule.cpp
+++ b/src/JITModule.cpp
@@ -321,7 +321,7 @@ void JITModule::compile_module(std::unique_ptr<llvm::Module> m, const string &fu
     std::vector<llvm::JITEventListener *> listeners;

     if (target.arch == Target::X86) {
-        listeners.push_back(llvm::JITEventListener::createIntelJITEventListener());
+        // listeners.push_back(llvm::JITEventListener::createIntelJITEventListener());
     }
     // TODO: If this ever works in LLVM, this would allow profiling of JIT code with symbols with oprofile.
     //listeners.push_back(llvm::createOProfileJITEventListener());

